samples/GenericHostSample/ServiceBaseLifetime.cs

```diff
@@ -23,23 +23,25 @@ public static Task RunAsServiceAsync(this IHostBuilder hostBuilder, Cancellation
 
     public class ServiceBaseLifetime : ServiceBase, IHostLifetime
     {
-        private Action<object> _startCallback;
-        private Action<object> _stopCallback;
+        private Func<object, Task> _startCallback;
         private object _startState;
-        private object _stopState;
 
-        public void RegisterDelayStartCallback(Action<object> callback, object state)
+        public ServiceBaseLifetime(IApplicationLifetime applicationLifetime)
+        {
+            ApplicationLifetime = applicationLifetime ?? throw new ArgumentNullException(nameof(applicationLifetime));
+        }
+
+        private IApplicationLifetime ApplicationLifetime { get; }
+
+        public Task RegisterDelayStartCallbackAsync(Func<object, Task> callback, object state)
         {
             _startCallback = callback ?? throw new ArgumentNullException(nameof(callback));
             _startState = state;
 
-            Run(this);
-        }
+            ApplicationLifetime.ApplicationStopping.Register(Stop);
 
-        public void RegisterStopCallback(Action<object> callback, object state)
-        {
-            _stopCallback = callback ?? throw new ArgumentNullException(nameof(callback));
-            _stopState = state;
+            Task.Run(() => Run(this)); // Otherwise this would block and prevent Start from finishing.
```

Comment:
Spawn a real thread so we don't block a thread pool thread.

Reply:
I uncovered a design flaw when testing this. RegisterDelayStartCallbackAsync has no abort mechanic. If you try to run your service app outside of a service Run(this) pops up a dialog box to tell you what you did wrong, but Run itself doesn't throw, it exits early. OnStart is never called so the delay start callback is never called, and the app hangs in StartAsync.